### PR TITLE
Make Users optional arguments

### PR DESCRIPTION
**Pull request by simpleauthority**
This PR allows `User`s to be optional command arguments similar to how the annotation operates on Bukkit's `Player`.

### CODE REVIEW

File: jda/src/main/java/co/aikar/commands/JDACommandContexts.java [hunk 1]
```diff
@@ -66,7 +66,11 @@ public JDACommandContexts(JDACommandManager manager) {
             if (c.hasAnnotation(SelfUser.class)) {
                 return jda.getSelfUser();
             }
-            String arg = c.popFirstArg(); // we pop because we are only issuer aware if we are annotated
+            String arg = c.getFirstArg();
+            if (c.hasAnnotation(Optional.class) && (arg == null || arg.isEmpty())) {
```
Comment (aikar) [inline jda/src/main/java/co/aikar/commands/JDACommandContexts.java]: This should be c.isOptional()

Comment (simpleauthority) [reply jda/src/main/java/co/aikar/commands/JDACommandContexts.java]: Oh is the annotation processed in core?

Comment (simpleauthority) [reply jda/src/main/java/co/aikar/commands/JDACommandContexts.java]: In Bukkit, it's checking against whether the context has the annotation.

Comment (aikar) [reply jda/src/main/java/co/aikar/commands/JDACommandContexts.java]: Feel free to update the other refs for me then :)

The goal is, that only the init process cares about actual annotations, which is ran through processors. 

then runtime just checks parsed results from the initialization.

it avoids reflection api's during runtime.

also, the annotation processor may make a custom annotation define something as Optional w/o it using the Optional annotation. so for accuracy, it NEEDS c.isOptional()

Comment (simpleauthority) [reply jda/src/main/java/co/aikar/commands/JDACommandContexts.java]: Sounds good, I'll edit soon

**Comment by aikar** [comment]
@simpleauthority did you forget about this

**Comment by simpleauthority** [comment]
@aikar yes, Checking it out now.

**Comment by simpleauthority** [comment]
Accidentally didn't configure muh git so buildtools is a contributor now

**Comment by simpleauthority** [comment]
@aikar dis done?

**Pull request merged**